implement js-doc-insert-function-doc-snippet

blaenk opened the pull request:
This introduces `js-doc-insert-function-doc-snippet`.

This uses yasnippet (if it's available) to insert the jsdoc comment which enables the use of placeholder values and cycling through the various fields.

To enable this, common functionality has been factored out to separate functions, most notably `js-doc--function-doc-metadata` which parses the function metadata into an alist which is then used by the jsdoc insertion functions.

You can see the advantages of the new `M-x js-doc-insert-function-doc-snippet` with this recording:

[![recording](https://i.imgur.com/333SwwC.gif)](https://i.imgur.com/333SwwC.gifv)

While I was at it I changed the regular `js-doc-insert-function-doc` to save the position of the jsdoc description line so that the point/cursor is moved to it after the jsdoc is inserted (see #13).

This supersedes #14.

blaenk committed:
implement js-doc-insert-function-doc-snippet

This uses yasnippet to insert the jsdoc comment which enables the use of
placeholder values and cycling through the various fields.

To enable this, common functionality has been factored out to separate
functions, most notably js-doc--function-doc-metadata which parses the
function metadata into an alist which is then used by the jsdoc
insertion functions.

futurist commented:
Using Mac10.9 + Emacs 25.0.95, with js2-mode enabled, the point is inside `random` function, the result is below:

![2016-07-13 22 27 44](https://cloud.githubusercontent.com/assets/159167/16806955/35cd254e-4949-11e6-8ba3-089663494771.png)


The tested code snippet as below:
```js
  var reAtRule = /^\s*@/
  var reClass = /:global\s*\(\s*((?:\.[A-Za-z0-9_-]+\s*)+)\s*\)|(\.)([!A-Za-z0-9_-]+)/g

// default local prefix implement
var random = (function () {
  var count = 0
  return function () {
    count++
    return '_' + Math.floor(Math.random() * Math.pow(2, 32)).toString(36) + count + '_'
  }
})()

```

The Emacs 25 have many weird behavior under Mac, tomorrow I'll test in windows, maybe it's more accurate.

blaenk commented:
Actually I think it's a simple bug. I'll take a look at it, thanks for trying it and reporting this.

blaenk commented:
Did jsdoc previously work with that Immediately Invoked Function Expression (IIFE)? Just to be sure.

blaenk committed:
fix positioning

The insertion functions are responsible for positioning the point at the
beginning of the function by calling js-doc--beginning-of-defun, which
is necessary for js-doc--function-doc-metadata to work.

The position-compensation after the point moved around for parsing the
data that was performed with the previous implementation is no longer
necessary because the metadata parsing is done in a separate excursion
via save-excursion.

blaenk commented:
@futurist Please try the latest commit. It now works with the two functions in your sample code. If you find any other problems please let me know.

[![recording](https://i.imgur.com/NnFL60V.gif)](https://i.imgur.com/NnFL60V.gifv)

mooz commented:
Great! Thx for the contribution.
